File: day9/haskell/day9.hs
solve :: Bool -> Int -> Int -> Int -> String -> (Int, Int)
solve _ _ s n []    = (s, n)
solve g d s n (c:xs)
    | c == '!'      = solve g d s n (tail xs)
    | c /= '>' && g = solve g d s (n + 1) xs
    | c == '<'      = solve True d s n xs
    | c == '>'      = solve False d s n xs
    | c == '{'      = solve g (d + 1) (s + d + 1) n xs
    | c == '}'      = solve g (d - 1) s n xs
    | otherwise     = solve g d s n xs

main :: IO ()
main = do
    input <- readFile "day9/input.txt"
    print . solve False 0 0 0 $ input

File: day9/typescript/day9.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

const solve = (chars: string[]) => {
    let garbage = false;
    let depth = 0;
    let sum = 0;
    let count = 0;
    while (chars.length > 0) {
        const c = chars.pop();
        switch(c) {
            case '!':
                chars.pop();
                break;
            case '<':
                garbage ? count++ : garbage = true;
                break;
            case '>':
                garbage = false;
                break;
            case '{':
                garbage ? count++ : sum += ++depth;
                break;
            case '}':
                garbage ? count++ : depth--;
                break;
            default:
                if (garbage) count++;
                break;
        }
    }

    return [sum, count];
}

(async () => {
    const input = await promisify(readFile)('day9/input.txt', 'utf8');
    const chars = Array.from(input).reverse();

    const [part1, part2] = solve(chars);
    console.log(part1);
    console.log(part2);
})();